File: ProblemSolve/104.cpp
/********************************
 * Maximum Depth of Binary Tree *
 ********************************/

class Solution {
public:
    //BFS
    int maxDepth(TreeNode* root) {        
        queue<pair<TreeNode*,int>> q;
        
        if(root == NULL){
            return 0;
        }
        
        q.push(make_pair(root, 1));
        TreeNode* node;
        int depth;
        int ans=0;
        
        while(!q.empty()){
            node = q.front().first;
            depth = q.front().second;
            q.pop();
            
            if(node -> left != NULL){
                q.push(make_pair(node->left, depth+1));
            }
            
            if(node->right != NULL){
                q.push(make_pair(node->right, depth+1));
            }
            
            if(node -> left == NULL && node->right == NULL){
                ans = max(ans, depth);
            }
        }
        
        return ans;
    }
    
    //DFS
    int maxDepth(TreeNode* root){
        if(root == NULL){
            return 0;
        }
        
        int leftDepth = maxDepth(root -> left);
        int rightDepth = maxDepth(root -> right);
        
        int maximumDepth = max(leftDepth, rightDepth) + 1;
        
        return maximumDepth;        
    }
};

File: ProblemSolve/104.cs
/********************************
 * Maximum Depth of Binary Tree *
 ********************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    int depth = 0;
    void Traverse(TreeNode root, int currDepth){
        if(root == null){
            return;
        }
        
        Traverse(root.left, currDepth + 1);
        Traverse(root.right, currDepth + 1);
        
        depth = Math.Max(depth, currDepth);                
    }
    
    public int MaxDepth(TreeNode root) {
        Traverse(root, 1);
        return depth;
    }
}